File: backend/src/services/universeRegistry.ts
import * as fs from 'fs';
import * as fsp from 'fs/promises';
import * as path from 'path';
import { DatabaseSync } from 'node:sqlite';

type UniverseSchema = 'symbols' | 'optionable' | 'source_symbols' | 'stocks';

type UniverseEntry = {
  path: string;
  schema: UniverseSchema;
  description?: string;
};

type UniverseRegistryFile = {
  generated_at?: string;
  universes?: Record<string, UniverseEntry>;
};

const REGISTRY_PATH = path.join(__dirname, '../../data/universe/registry.json');
const STOCK_EXCLUSIONS_PATH = path.join(__dirname, '../../data/universe/stock_exclusions.json');
const SYMBOL_CATALOG_DB_PATH = path.join(__dirname, '../../data/symbol-catalog.sqlite');
const CATALOG_UNIVERSES = new Set(['tradable_stock_default', 'tradable_optionable_stocks']);

function loadCatalogUniverseSymbolsSync(name: string): string[] {
  if (!CATALOG_UNIVERSES.has(String(name || '').trim())) return [];
  if (!fs.existsSync(SYMBOL_CATALOG_DB_PATH)) return [];
  let db: DatabaseSync | null = null;
  try {
    db = new DatabaseSync(SYMBOL_CATALOG_DB_PATH, { readOnly: true });
    const where = [
      "s.asset_class = 'stocks'",
      "m.membership_type = 'eligibility'",
      "m.membership_value = 'tradable_stock_default'",
    ];
    if (name === 'tradable_optionable_stocks') {
      where.push('s.optionable = 1');
    }
    const rows = db.prepare(`
      SELECT s.symbol
      FROM symbols s
      JOIN symbol_memberships m
        ON m.symbol = s.symbol
      WHERE ${where.join(' AND ')}
      ORDER BY s.symbol ASC
    `).all() as Array<{ symbol?: string }>;
    return normalizeSymbols(rows.map((row) => row?.symbol));
  } catch {
    return [];
  } finally {
    try { db?.close(); } catch {}
  }
}

function normalizeSymbols(values: unknown): string[] {
  if (!Array.isArray(values)) return [];
  const out: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    const symbol = String(value || '').trim().toUpperCase();
    if (!symbol || seen.has(symbol)) continue;
    seen.add(symbol);
    out.push(symbol);
  }
  return out;
}

function applyStockExclusions(symbols: string[], excluded: Set<string>): string[] {
  if (!excluded.size) return symbols;
  return symbols.filter((symbol) => !excluded.has(symbol));
}

function parsePayloadSymbols(payload: any, schema: UniverseSchema): string[] {
  if (schema === 'stocks') {
    const stocks = Array.isArray(payload?.stocks) ? payload.stocks : [];
    return normalizeSymbols(stocks.map((item: any) => item?.ticker));
  }
  if (schema === 'optionable') {
    return normalizeSymbols(payload?.optionable || payload?.symbols || []);
  }
  if (schema === 'source_symbols') {
    return normalizeSymbols(payload?.source_symbols || payload?.symbols || []);
  }
  return normalizeSymbols(payload?.symbols || payload);
}

function readRegistrySync(): UniverseRegistryFile {
  try {
    return JSON.parse(fs.readFileSync(REGISTRY_PATH, 'utf-8')) as UniverseRegistryFile;
  } catch {
    return {};
  }
}

function readExcludedSymbolsSync(): Set<string> {
  try {
    const payload = JSON.parse(fs.readFileSync(STOCK_EXCLUSIONS_PATH, 'utf-8'));
    return new Set(normalizeSymbols(payload?.symbols || payload));
  } catch {
    return new Set();
  }
}

async function readRegistry(): Promise<UniverseRegistryFile> {
  try {
    return JSON.parse(await fsp.readFile(REGISTRY_PATH, 'utf-8')) as UniverseRegistryFile;
  } catch {
    return {};
  }
}

async function readExcludedSymbols(): Promise<Set<string>> {
  try {
    const payload = JSON.parse(await fsp.readFile(STOCK_EXCLUSIONS_PATH, 'utf-8'));
    return new Set(normalizeSymbols(payload?.symbols || payload));
  } catch {
    return new Set();
  }
}

function resolveUniversePath(entry: UniverseEntry): string {
  return path.resolve(path.dirname(REGISTRY_PATH), entry.path);
}

function loadEntrySymbolsSync(entry: UniverseEntry): string[] {
  try {
    const payload = JSON.parse(fs.readFileSync(resolveUniversePath(entry), 'utf-8'));
    return parsePayloadSymbols(payload, entry.schema);
  } catch {
    return [];
  }
}

async function loadEntrySymbols(entry: UniverseEntry): Promise<string[]> {
  try {
    const payload = JSON.parse(await fsp.readFile(resolveUniversePath(entry), 'utf-8'));
    return parsePayloadSymbols(payload, entry.schema);
  } catch {
    return [];
  }
}

export function loadUniverseSymbolsSync(name: string): string[] {
  if (CATALOG_UNIVERSES.has(name)) {
    return applyStockExclusions(loadCatalogUniverseSymbolsSync(name), readExcludedSymbolsSync());
  }
  const registry = readRegistrySync();
  const entry = registry?.universes?.[name];
  if (!entry) return [];
  return applyStockExclusions(loadEntrySymbolsSync(entry), readExcludedSymbolsSync());
}

export async function loadUniverseSymbols(name: string): Promise<string[]> {
  if (CATALOG_UNIVERSES.has(name)) {
    return applyStockExclusions(loadCatalogUniverseSymbolsSync(name), await readExcludedSymbols());
  }
  const registry = await readRegistry();
  const entry = registry?.universes?.[name];
  if (!entry) return [];
  return applyStockExclusions(await loadEntrySymbols(entry), await readExcludedSymbols());
}

File: backend/services/universe_registry.py
from __future__ import annotations

import json
import os
from functools import lru_cache
from typing import Any, Dict, List, Set


DATA_DIR = os.path.normpath(os.path.join(os.path.dirname(__file__), "..", "data"))
UNIVERSE_DIR = os.path.join(DATA_DIR, "universe")
REGISTRY_PATH = os.path.join(UNIVERSE_DIR, "registry.json")
STOCK_EXCLUSIONS_PATH = os.path.join(UNIVERSE_DIR, "stock_exclusions.json")
MARKET_CAP_SNAPSHOT_PATH = os.path.join(DATA_DIR, "market_cap_snapshot.json")


def _normalize_symbols(values: Any) -> List[str]:
    if not isinstance(values, list):
        return []
    out: List[str] = []
    seen: Set[str] = set()
    for value in values:
        symbol = str(value or "").strip().upper()
        if not symbol or symbol in seen:
            continue
        seen.add(symbol)
        out.append(symbol)
    return out


@lru_cache(maxsize=1)
def load_stock_exclusions() -> Set[str]:
    try:
        with open(STOCK_EXCLUSIONS_PATH, "r", encoding="utf-8") as f:
            payload = json.load(f)
        values = payload.get("symbols") if isinstance(payload, dict) else payload
        return set(_normalize_symbols(values))
    except Exception:
        return set()


def _apply_stock_exclusions(symbols: List[str]) -> List[str]:
    excluded = load_stock_exclusions()
    if not excluded:
        return symbols
    return [symbol for symbol in symbols if symbol not in excluded]


@lru_cache(maxsize=1)
def _load_registry() -> Dict[str, Any]:
    try:
        with open(REGISTRY_PATH, "r", encoding="utf-8") as f:
            return json.load(f) or {}
    except Exception:
        return {}


def resolve_universe_path(name: str) -> str | None:
    registry = _load_registry()
    entry = ((registry.get("universes") or {}) if isinstance(registry, dict) else {}).get(name) or {}
    rel_path = str(entry.get("path") or "").strip()
    if not rel_path:
        return None
    return os.path.normpath(os.path.join(UNIVERSE_DIR, rel_path))


def _parse_payload_symbols(payload: Any, schema: str) -> List[str]:
    if schema == "stocks":
        stocks = payload.get("stocks") if isinstance(payload, dict) else []
        return _normalize_symbols([
            item.get("ticker") for item in stocks
            if isinstance(item, dict)
        ])
    if schema == "optionable":
        return _normalize_symbols((payload or {}).get("optionable") or (payload or {}).get("symbols") or [])
    if schema == "source_symbols":
        return _normalize_symbols((payload or {}).get("source_symbols") or (payload or {}).get("symbols") or [])
    if isinstance(payload, dict):
        return _normalize_symbols(payload.get("symbols") or [])
    return _normalize_symbols(payload)


def load_universe_symbols(name: str) -> List[str]:
    registry = _load_registry()
    entry = ((registry.get("universes") or {}) if isinstance(registry, dict) else {}).get(name) or {}
    schema = str(entry.get("schema") or "symbols").strip()
    abs_path = resolve_universe_path(name)
    if not abs_path:
        return []
    try:
        with open(abs_path, "r", encoding="utf-8") as f:
            payload = json.load(f)
        return _apply_stock_exclusions(_parse_payload_symbols(payload, schema))
    except Exception:
        return []


def load_clean_stock_symbols() -> List[str]:
    return load_universe_symbols("clean_stocks")


@lru_cache(maxsize=1)
def load_market_cap_snapshot_billions() -> Dict[str, float]:
    try:
        with open(MARKET_CAP_SNAPSHOT_PATH, "r", encoding="utf-8") as f:
            payload = json.load(f) or {}
        out: Dict[str, float] = {}
        for symbol, raw_value in payload.items():
            sym = str(symbol or "").strip().upper()
            if not sym:
                continue
            try:
                value = float(raw_value)
            except (TypeError, ValueError):
                continue
            if value > 0:
                out[sym] = value / 1e9
        return out
    except Exception:
        return {}

